### PR TITLE
[Model][Perf] Optimize LTX Ulysses sequence parallelism

### PR DESCRIPTION
**Pull request by mglyn**
## Summary

- Keep LTX's short audio and prompt streams replicated while sharding the long video sequence under Ulysses.
- Add an LTX-specific video-to-audio attention path that avoids redundant collectives and fails closed for unsupported SP layouts.
- Disable LTX's one-frame startup dummy run: it does not warm up the production video shape and adds substantial startup time with default compilation. MiniMax-H3 already disables the same generic warmup.
- Expose Stage 1, Stage 2, and VAE decode profiler timings, and document the recommended multi-GPU low-latency recipe.

## Performance

NVIDIA H200, LTX-2.5 distilled two-stage, 1920x1088, 121 frames, default compile path (`--enforce-eager` unset), and production automatic attention routing (`FLASH_ATTN`). Both sides skip the one-frame startup dummy warmup. Each result uses one explicit full-shape warmup followed by the mean of three measured requests. Lower is better.

The baseline is `origin/main@0e903e969` with identical profiler instrumentation and startup warmup policy, so the measured delta isolates the Ulysses change.

`stage_0_gen_ms` is Omni's aggregate generation wall time for the single diffusion stage;

| Variant | Denoise 1 (ms) | Denoise 2 (ms) | Decode (ms) | `stage0_gen` (ms) | `delta_ms` |
| --- | ---: | ---: | ---: | ---: | ---: |
| `main ulysses4` | 1,613.4 | 2,503.6 | 3,758.2 | 8,854.3 | — |
| `current ulysses4` | 1,438.9 | 2,440.5 | 3,758.4 | 8,647.5 | **-206.8 (-2.34%)** |
| `main ulysses8` | 1,802.0 | 1,332.8 | 2,181.7 | 6,582.4 | — |
| `current ulysses8` | 1,338.3 | 1,286.4 | 2,189.3 | 5,809.5 | **-772.9 (-11.74%)** |

**Observation:** As the Ulysses rank count increases, compute per rank decreases while communication becomes a larger share of denoising time, so eliminating redundant collectives has a larger impact. This is reflected in the `stage0_gen` improvement growing from 2.34% at 4 ranks to 11.74% at 8 ranks, and the Stage 1 denoise improvement growing from 10.81% to 25.73%.

## Video Demo

[](https://github.com/user-attachments/assets/1a664c79-663d-41d4-b24e-44cd795eea51)

The optimized SP communication layout changes BF16 attention execution ordering and therefore floating-point rounding. Diffusion amplifies this per-step drift, so the outputs are not pixel-equivalent even with the same prompt, seed, and configuration.

On the four-GPU side-by-side demo, current vs. main measures **22.11 dB PSNR** over all 121 raw decoded RGB uint8 frames at 1088x1920, before cropping, labels, or MP4 encoding. The comparison is intended as a visual-quality sanity check rather than a claim of numerical equivalence.

Demo artifact: 2160x1920 side-by-side (each 1080x1920 panel is center-cropped from the 1088x1920 raw output), 121 frames at 24 fps (5.04 s), browser-compatible H.264/yuv420p with AAC audio from main, faststart, 5.83 MB.

### Acknowledgements and licensing

Thanks to Lightricks for publishing LTX-2 and documenting its [sequence-parallel design](https://github.com/Lightricks/LTX-2/blob/main/packages/ltx-pipelines/docs/multigpu/sequence-parallel.md).
The LTX2 model integration is derived from Hugging Face Diffusers under Apache-2.0, and this change builds on vLLM-Omni's existing Apache-2.0 Ulysses and joint-attention primitives.

vLLM-Omni's Apache-2.0 license covers this implementation. LTX model weights and separately licensed Lightricks artifacts remain subject to the [applicable LTX license](https://github.com/Lightricks/LTX-2/blob/main/LICENSE).

## Tests

- `pytest -q tests/diffusion/models/ltx2/test_ltx2_sequence_parallel.py tests/diffusion/models/ltx2/test_ltx2_transformer.py tests/diffusion/models/ltx2/test_ltx25_pipeline.py` (77 passed)
- `pytest -q tests/diffusion/models/ltx2/test_ltx2_pipeline.py tests/diffusion/test_diffusion_engine_dummy_run.py` (143 passed)
- Ruff and `git diff --check`

### CODE REVIEW

File: tests/diffusion/models/ltx2/test_ltx2_sequence_parallel.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            key = global_key.chunk(2, dim=1)[ulysses_rank].contiguous()
+            value = global_value.chunk(2, dim=1)[ulysses_rank].contiguous()
+            mask = None
+            if masked:
+                mask = torch.ones(batch_size, 1, 1, 6, dtype=torch.bool)
+                mask[0, ..., -1] = False
+                mask[1, ..., -2:] = False
+            metadata = AttentionMetadata(attn_mask=mask) if masked else None
+
+            local_query, local_key, local_value, local_metadata, ctx = strategy.pre_attention(
```
Comment (david6666666) [inline tests/diffusion/models/ltx2/test_ltx2_sequence_parallel.py:64]: Could we add a tiny transformer parity test using the real `_sp_plan` hooks and attention dispatch, comparing both video and audio outputs against SP=1? This calls the strategy directly, so regressions in mode routing or RoPE/timestep sharding would still pass.

**Review by david6666666 — COMMENTED** [review]
The remaining validation gap is parity through the full transformer path.

**Comment by david6666666** [comment]
Please provide the link to the official LTX reference design.

File: tests/diffusion/models/ltx2/test_ltx2_sequence_parallel.py (hunk 1 above)
Comment (mglyn) [reply tests/diffusion/models/ltx2/test_ltx2_sequence_parallel.py:64]: Added in 54b3a840f. The new `test_ltx2_transformer_ulysses.py` builds a tiny one-block `LTX2VideoTransformer3DModel`, runs the SP=1 reference and Ulysses=2 through the real `_sp_plan` hooks and attention dispatch, and compares both video and audio transformer outputs.

The local 2-GPU check on NVIDIA H200 passed (`1 passed in 35.19s`); the observed relative L2 was `0.0` for both outputs on both ranks. The test comment also makes its boundary explicit: this FP32/SDPA check validates transformer mathematical-path parity, but does not claim pixel-level E2E parity under BF16/compiled attention, where rounding drift can be amplified across diffusion steps.

**Comment by mglyn** [comment]
Thanks — the official LTX sequence-parallel reference design is documented here: https://github.com/Lightricks/LTX-2/blob/main/packages/ltx-pipelines/docs/multigpu/sequence-parallel.md. The PR description also links to it under **Acknowledgements and licensing**.

**Review by david6666666 — APPROVED** [review]
LGTM now
